refactor!: Using the MsgEthereumTx interface in the rpc module

nowooj opened the pull request:
# Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review. -->

<!-- Please keep your PR as draft until it's ready for review -->

<!-- Pull requests that sit inactive for longer than 30 days will be closed.  -->
This code doesn't cover all MsgEthereumTx implementations within the rpc module, but it does cover the query part. This code can be expanded and structured into a completely independent code structure.

[Removing Hash()](https://github.com/cosmos/evm/commit/6d023bb4002886584197a3ddaa35590a21134eec) hasn't important to do with the PR, but it was added for backward compatibility since ethermint uses Hash as a variable.

Closes: #853

---

## Author Checklist

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.

I have...

- [ ] tackled an existing issue or discussed with a team member
- [ ] left instructions on how to review the changes
- [x] targeted the `main` branch

nowooj committed:
refactor: rpc use interface MsgEthereumTx

vladjdk commented on ante/tx_listener.go:
```diff
@@ -27,7 +27,7 @@ func (d TxListenerDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate boo
 	if ctx.IsCheckTx() && !simulate && d.pendingTxListener != nil {
 		for _, msg := range tx.GetMsgs() {
 			if ethTx, ok := msg.(*evmtypes.MsgEthereumTx); ok {
```
Why are we leaving some *evmtypes.MsgEthereumTx as pointers instead of the interface?

nowooj committed:
refactor: remove Hash() func for backward compatibility

fix: remove Hash() func for backward compatibility

vladjdk reviewed (changes requested):
Lint and conflicts

vladjdk commented:
@greptile review

nowooj committed:
fix: unsafe slice-to-array conversion

nowooj committed:
refactor: inverted dependecy

nowooj commented on rpc/backend/comet_to_eth.go:
```diff
@@ -292,8 +294,8 @@ func (b *Backend) ReceiptsFromCometBlock(
-		if ethMsg.Raw.To() == nil {
-			contractAddress = crypto.CreateAddress(ethMsg.GetSender(), ethMsg.Raw.Nonce())
+		if tx.To() == nil {
+			contractAddress = crypto.CreateAddress(common.Address(ethMsg.GetFrom().Bytes()), tx.Nonce())
```
https://github.com/cosmos/evm/pull/854/commits/cd50ea2f5bceaf07d79a6fff36180f851ddb78e2

nowooj commented on x/vm/types/msg.go:
```diff
@@ -13,6 +13,7 @@ import (
 	protov2 "google.golang.org/protobuf/proto"
 
 	evmapi "github.com/cosmos/evm/api/cosmos/evm/vm/v1"
+	"github.com/cosmos/evm/rpc/types/interfaces"
```
https://github.com/cosmos/evm/pull/854/commits/56544ac8bdf1099dc477f5bfe2150f7f99ef7efb

nowooj commented on rpc/types/interfaces/interface.go:
```diff
@@ -0,0 +1,22 @@
+	GetSenderLegacy(signer ethtypes.Signer) (common.Address, error)
+	AsMessage(baseFee *big.Int) *core.Message
+	UnmarshalBinary(b []byte, signer ethtypes.Signer) error
+}
```
https://github.com/cosmos/evm/pull/854/commits/40ace1edeab01bbf304c68f829efa684d09f53d4

nowooj committed:
refactor: update interface to RPCMsgEthereumTxI for evm rpc's MsgEthereumTx

aljo242 reviewed (changes requested):
**ci has not run:** fork pr, all workflows `action_required`. also 34 commits behind main.

**`Hash()` removal is a bundled breaking change:** the pr description acknowledges this was kept for ethermint backward compat. removing it breaks any downstream calling `.Hash()` directly on `*MsgEthereumTx`. move this to a separate pr with an explicit changelog entry.

**undocumented behavioral change in contract address derivation:** `comet_to_eth.go:299` changes `ethMsg.GetSender()` (recovered from signature) to `ethMsg.GetFrom()` (stored field). these should be equivalent for well-formed txs, but the difference isn't documented and could silently produce wrong contract addresses for edge cases. needs a comment or test confirming equivalence.

**`GetSenderLegacy` in interface but unused:** no call site in this diff uses it through the interface -- either document why it's required or remove it.

nowooj committed:
Merge branch 'main' into refactor/interface-rpcEthTx

nowooj committed:
fix: sender address calculation in ReceiptsFromCometBlock

nowooj committed:
chore: update CHANGELOG to reflect API-breaking changes

nowooj commented:
> **ci has not run:** fork pr, all workflows `action_required`. also 34 commits behind main.
> 
> **`Hash()` removal is a bundled breaking change:** the pr description acknowledges this was kept for ethermint backward compat. removing it breaks any downstream calling `.Hash()` directly on `*MsgEthereumTx`. move this to a separate pr with an explicit changelog entry.
> 
> **undocumented behavioral change in contract address derivation:** `comet_to_eth.go:299` changes `ethMsg.GetSender()` (recovered from signature) to `ethMsg.GetFrom()` (stored field). these should be equivalent for well-formed txs, but the difference isn't documented and could silently produce wrong contract addresses for edge cases. needs a comment or test confirming equivalence.
> 
> **`GetSenderLegacy` in interface but unused:** no call site in this diff uses it through the interface -- either document why it's required or remove it.

@aljo242 Thanks, addressed the actionable parts.

- Added an `API-BREAKING` changelog entry for the `MsgEthereumTx.Hash()` removal and documented `AsTransaction().Hash()` as the replacement.
- Updated contract creation receipt address derivation to use `GetSenderLegacy(signer)` instead of `GetFrom()` directly, preserving the fallback sender recovery for older decoded txs with an empty `From`.
- `GetSenderLegacy` is used through `RPCMsgEthereumTxI` in RPC/query paths (`getAccountNonce`, `GetTransactionReceipt`, `GetBlockReceipts`), so I kept it in the interface.
- The branch has also been updated with latest `main`.

nowooj committed:
refactor: remove GetFrom from RPC tx interface

nowooj committed:
Merge branch 'main' into refactor/interface-rpcEthTx

nowooj committed:
Merge branch 'main' into refactor/interface-rpcEthTx

nowooj committed:
Merge branch 'main' into refactor/interface-rpcEthTx

nowooj committed:
Merge branch 'main' into refactor/interface-rpcEthTx
